### ISSUE TITLE
Cannot validate the overlap detection model

### ISSUE THREAD

**Issue by yining4869** [pyannote/pyannote-audio#376]
Hi,
I have trained the overlap detection model then I want to validate that and I just followed tutorial. But there are some errors and I cannot fix it. Can u help me to have a check?

  File "/home/yininggao/anaconda3/envs/pyannote/bin/pyannote-audio", line 8, in <module>
    sys.exit(main())
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/pyannote_audio.py", line 380, in main
    app.validate(protocol, **params)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/base.py", line 488, in validate
    **kwargs)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/overlap_detection.py", line 133, in validate_epoch
    _ = self.pool_.map(validate, validation_data)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/multiprocessing/pool.py", line 268, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/multiprocessing/pool.py", line 657, in get
    raise self._value
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/multiprocessing/pool.py", line 431, in _handle_tasks
    put(task)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: can't pickle _thread.RLock objects

**Comment by yining4869** [comment]
And I found that I didn't update to the latest version. But after I update to the latest version, the validate process didn't start and always remained the same status for a long time(about 10 minutes)as below:
0iteration [00:00, ?iteration/s]
When I interrupt the process with keyboard, it shows the programme are stuck in the part as below:
  File "/home/yininggao/anaconda3/envs/pyannote/bin/pyannote-audio", line 8, in <module>
    sys.exit(main())
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/pyannote_audio.py", line 422, in main
    app.validate(protocol, **params)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/base.py", line 358, in validate
    **kwargs,
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/applications/speech_detection.py", line 77, in validate_epoch
    current_file["scores"] = pretrained(current_file)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/features/base.py", line 152, in __call__
    features = self.get_features(y.data, sample_rate)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/features/pretrained.py", line 207, in get_features
    progress_hook=self.progress_hook,
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/train/model.py", line 468, in slide
    tfX = self(tX, return_intermediate=return_intermediate)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/models/models.py", line 543, in forward
    output = self.rnn_(output)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/pyannote/audio/models/models.py", line 217, in forward
    output, hidden = self.rnn_(features)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/yininggao/anaconda3/envs/pyannote/lib/python3.7/site-packages/torch/nn/modules/rnn.py", line 559, in forward
    self.dropout, self.training, self.bidirectional, self.batch_first)
 Do u know why does this happen? Looking forward to your reply and thanks a lot!

**Comment by hbredin** [comment]
Looks like it is indeed validating. Just wait a bit longer :-)

Are you running the validation on GPU? 
How big is your validation set?

**Comment by yining4869** [comment]
Yes, I found that it is actually validating but with a very very slow speed. And I'm running the validation on GPU, the validation set is about 2GB. And I also use the --parallel option, try to speed up the validate speed but there's no significant improvement.
Should I just cut the size of validation set to speed up?

**Comment by yining4869** [comment]
> Looks like it is indeed validating. Just wait a bit longer :-)
> 
> Are you running the validation on GPU?
> How big is your validation set?

Thanks for your quick reply. 
Yes, I found that it is actually validating but with a very very slow speed. And I'm running the validation on GPU, the validation set is about 2GB. And I also use the --parallel option, try to speed up the validate speed but there's no significant improvement.
Should I just cut the size of validation set to speed up?

**Comment by hbredin** [comment]
`--parallel` option has no effect for validation. 

To speed things up, I suggest 
* you use the `--every=N` option to only validate every N epochs
* you reduce the validation size down to only a few hours of audio.

**Comment by hbredin** [comment]
Closing as I think the original issue has been solved. Please re-open if needed.

**Comment by yining4869** [comment]
Hi, sorry for the late reply.
I have reduced the validation size to just a few hours and used --every=N to only validate every N epochs. The command line is shown below:
**pyannote-audio ovl validate --gpu --subset=development --from=300 --to=900 --every=200 ${TRN_DIR} AMI.SpeakerDiarization.MixHeadset**
But it is still remained the same status for a long time(I have waited for almost 30 minutes):
**0iteration [00:00, ?iteration/s]**
Meanwhile,  at the beginning of validating, the GPU-Util kept changing normally and the Memory-Usage was 35%.But after a few minutes, the GPU-Util became 0 and the Memory-Usage was still 35%. 
Do u know why does this happen? Looking forward to your reply and thanks a lot!
